File: 2026-01/src/day_02/boj_28279/Main.cpp
#include <bits/stdc++.h>
using namespace std;

const int MAX = 2000000;
int dq[MAX];
int s = MAX / 2, e = MAX / 2;

void push_front(int x) {
    dq[--s] = x;
}

void push_back(int x) {
    dq[e++] = x;
}

void pop_front() {
    s++;
}

void pop_back() {
    e--;
}

int size() {
    return e - s;
}

bool empty() {
    return s == e;
}

int front() {
    return dq[s];
}

int back() {
    return dq[e - 1];
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n;
    cin >> n;

    while (n--) {
        int q;
        cin >> q;

        if (q == 1) {
            int x;
            cin >> x;
            push_front(x);
        } else if (q == 2) {
            int x;
            cin >> x;
            push_back(x);
        } else if (q == 3) {
            if (empty()) {
                cout << -1 << '\n';
            } else {
                cout << front() << '\n';
                pop_front();
            }
        } else if (q == 4) {
            if (empty()) {
                cout << -1 << '\n';
            } else {
                cout << back() << '\n';
                pop_back();
            }
        } else if (q == 5) {
            cout << size() << '\n';
        } else if (q == 6) {
            cout << empty() << '\n';
        } else if (q == 7) {
            if (empty()) {
                cout << -1 << '\n';
            } else {
                cout << front() << '\n';
            }
        } else {
            if (empty()) {
                cout << -1 << '\n';
            } else {
                cout << back() << '\n';
            }
        }
    }
}

File: 2026-01/src/day_02/boj_28279/Main.java
package day_02.boj_28279;

import java.io.*;
import java.util.*;

public class Main {

    static final int MX = 2000000;
    static int[] deque = new int[MX];
    static int front = MX / 2, rear = MX / 2;

    static void push_front(int x) {
        deque[--front] = x;
    }

    static void push_back(int x) {
        deque[rear++] = x;
    }

    static int pop_front() {
        return deque[front++];
    }

    static int pop_back() {
        return deque[--rear];
    }

    static int size() {
        return rear - front;
    }

    static boolean empty() {
        return front == rear;
    }

    static int front() {
        return deque[front];
    }

    static int back() {
        return deque[rear - 1];
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        int n = Integer.parseInt(br.readLine());
        while (n-- > 0) {
            st = new StringTokenizer(br.readLine());
            int c = Integer.parseInt(st.nextToken());

            if (c == 1) {
                int x = Integer.parseInt(st.nextToken());
                push_front(x);
            } else if (c == 2) {
                int x = Integer.parseInt(st.nextToken());
                push_back(x);
            } else if (c == 3) {
                if (empty()) {
                    sb.append("-1\n");
                } else {
                    sb.append(pop_front()).append("\n");
                }
            } else if (c == 4) {
                if (empty()) {
                    sb.append("-1\n");
                } else {
                    sb.append(pop_back()).append("\n");
                }
            } else if (c == 5) {
                sb.append(size()).append("\n");
            } else if (c == 6) {
                if (empty()) {
                    sb.append("1\n");
                } else {
                    sb.append("0\n");
                }
            } else if (c == 7) {
                if (empty()) {
                    sb.append("-1\n");
                } else {
                    sb.append(front()).append("\n");
                }
            } else {
                if (empty()) {
                    sb.append("-1\n");
                } else {
                    sb.append(back()).append("\n");
                }
            }
        }

        System.out.println(sb);
    }
}
